akka-persistence-testkit/src/main/scala/akka/persistence/testkit/ChangePersisted.scala

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.testkit
+
+/** Common superclass denoting a persistence effect from an EventSourcedBehavior or a DurableStateBehavior.
+ *
+ *  From an EventSourcedBehavior, a ChangePersisted will be either an EventPersisted (corresponding to an
+ *  Effect.persist or Effect.persistAll) or a StatePersisted (corresponding to a snapshot being persisted).
+ *  An EventPersisted can have zero or more tags while a StatePersisted from an EventSourcedBehavior will
+ *  always have empty tags.
+ *
+ *  From a DurableStateBehavior, a ChangePersisted will always be a StatePersisted (corresponding to an
+ *  Effect.persist).  The StatePersisted may have a single tag.
+ *
+ *  EventPersisted can be distinguished from ChangePersisted by a type check (e.g. pattern match or instanceof)
+ *  or through the getEvent()/getState() methods: for an EventPersisted, getEvent() will be non-null and getState()
+ *  will be null, while for a StatePersisted, getEvent() will be null and getState() will be non-null.
+ */
+sealed abstract class ChangePersisted[State, +Event]() extends Product with Serializable {
```

Comment:
This feels a bit like cramming heterogenous types into a pretend-common-supertype, could we drop this or make it hidden and have the public APIs which already know if it is a value or es entity return the concrete types?

Reply:
could it be without the State and Event and then we or the user can pattern match (java instanceof) to get to the concrete details?

Reply:
I like @johanandren's idea of exposing distinct probes (more like `TestInbox` than `TestProbe`), which removes the need for the single channel.

My original implementation of this idea didn't even deal with snapshots in the event-sourced case so a single queue worked there: single queue with multiple types is inferior to multiple queues each with a single type.